Use macro for boilerplate code in webdriver `handle_*`

dlrobertson opened the pull request:
Add a macro for the boilerplate code that appears in most of Handler's methods handling a `WebDriverScriptCommand`.

There are at least three steps that appear in most `Hande::handle_*` methods that handle a `WebDriverScriptCommand`, so I created a simple macro that generates the equivalent code.

There are pros and cons to using a macro like this. It does reduce boilerplate code, but it also can make the code harder to read. Please let me know if you have any comments!

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9747)

<!-- Reviewable:end -->

danlrobertson commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[components/webdriver_server/lib.rs, line 72 \[r1\]](https://reviewable.io:443/reviews/servo/servo/9747#-KBLQCjKNA14jeuxjVCA:-KBLQCjKNA14jeuxjVCB:-431411545) ([raw file](https://github.com/servo/servo/blob/0b4ddd85f779c605d488d62cba52ce2f448618c7/components/webdriver_server/lib.rs#L72)):*
This is pretty ugly. I really don't like the two branches that basically do the same thing. Theoretically, this could be split into two macros. One than handles the input and one that does the work.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9747)*
<!-- Sent from Reviewable.io -->

frewsxcv commented:
Is there any reason you chose a macro besides variadic arguments? An inline function or method could maybe work.

e.g. instead of

```rust
script_command_handle!(self, sender, WebDriverScriptCommand::GetElementCSS, element.id.clone(), name.to_owned());
```

something like

```rust
script_command_handle(self, sender, WebDriverScriptCommand::GetElementCSS(element.id.clone(), name.to_owned()));
```

or maybe even

```rust
self.handle_script_command(sender, WebDriverScriptCommand::GetElementCSS(element.id.clone(), name.to_owned()));
```

frewsxcv commented:
Forgot to mention: Nice work! Glad to see all that duplicate code consolidated :sparkles:

KiChjang commented on components/webdriver_server/lib.rs:
```diff
@@ -55,6 +55,21 @@ use webdriver::response::{ElementRectResponse, NewSessionResponse, ValueResponse
+                                                   $t($i0)))
+                                      ).unwrap();
+    };
+    ($self_:ident, $i0:ident, $t:path, $($e:expr),*) => {
```
Shouldn't $i0 be named `$sender` instead?

danlrobertson commented:
> Is there any reason you chose a macro besides variadic arguments? An inline function or method could maybe work.

No reason :smile: I just read [The little book of rust macros](https://danielkeep.github.io/tlborm/book/README.html), so I guess I'm thinking about macros more. I'm 100% okay with switching this out for an `#[inline]` function. I just did it this way because I had an idea of how I'd implement it. Let me know which way is preferred and I'll make any necessary changes.

> Shouldn't `$i0` be named `$sender` instead?

Definitely. The update will reflect this.

jdm commented:
Yeah, I would prefer a method over a macro here. Thanks for tackling this, though!

danlrobertson commented:
Awesome! I'll make the changes and push them this evening

dlrobertson committed:
Macro for boilerplate code in webdriver handle_*

Add a macro for the boilerplate code that appears in most of Handler's
methods handling a WebDriverScriptCommand.

danlrobertson commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/webdriver_server/lib.rs, line 313 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9747#-KBQaT2LPZgB6-cYvn0v:-KBQaT2LPZgB6-cYvn0w:1046961336) ([raw file](https://github.com/servo/servo/blob/05547dce5a756a440ea6022665d1947dcd518f3f/components/webdriver_server/lib.rs#L313)):*
I totally forgot that some instances used `frame_pipeline` and others `root_pipeline` in the first two commits. Sorry about that

---

*[components/webdriver_server/lib.rs, line 352 \[r2\]](https://reviewable.io:443/reviews/servo/servo/9747#-KBQagaJI4ufzXkOmtd9:-KBQagaKGD0sin9IphLz:-757056601) ([raw file](https://github.com/servo/servo/blob/05547dce5a756a440ea6022665d1947dcd518f3f/components/webdriver_server/lib.rs#L352)):*
Kind of annoying that we have to `try!`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9747)*
<!-- Sent from Reviewable.io -->

Ms2ger commented:
@bors-servo r+

Thank you!

bors-servo commented:
:pushpin: Commit 05547dc has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger 05547dce5a756a440ea6022665d1947dcd518f3f -->

bors-servo commented:
:hourglass: Testing commit 05547dc with merge 6b81d13...

Ms2ger commented:
@bors-servo retry
* #7846

bors-servo commented:
:hourglass: Testing commit 05547dc with merge aaad24c...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1806), [gonk](http://build.servo.org/builders/gonk/builds/1847), [linux-dev](http://build.servo.org/builders/linux-dev/builds/1824), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2085), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/522), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/1877), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/1926)
